### DISCUSSION TITLE
Should we try Codespaces?

### DISCUSSION THREAD

**Discussion by mxie** [railsbridge/docs#701] [Ideas]
GitHub recently launched [Codespaces](https://github.com/features/codespaces), which could be a good alternative to native installations (as written in the curriculum currently), the VM setup (mostly used by the Boston chapter), or Cloud9.

What are some advantages and disadvantages of this approach?

**Comment by ultrasaurus** [comment]
One big disadvantages is that it's not clear how it works (which doesn't have to be a problem if we make, or find, answers to questions)...  

- what actions use "minutes" and in whose account (when pairing)?
  - if pairing in a codespace, does each person used up 1 minute per minute pairing?  or does the codespace owner get charged 1 minute or 2 minutes?   (personal 120 hours per month, from the [billing docs](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces) seems like the owner of the account is running a virtual machine and there is no extra charge for someone to login to it and run commands) -- we would want to explain about how someone knows how many "cores" they are using and make sure they are using a single core (unless there's a need for multiple cores, are there even single core machine types available?)
- what actions cause storage to be used? (personal 15GB free) For example, @macboypro and I were experimenting a few weeks ago and I ended up using <1GB of storage, but he did more configuration and used >7GB
  <img width="713" alt="image" src="https://user-images.githubusercontent.com/41906/231566999-c0b6f869-5c0b-4102-9606-c7c8ee0eb7f4.png">
- if I have a fork of someone else's repo and am working on a PR, whose codespace is it?  I would assume mine, but that conflicts with my experience where I was unable to set up a codespace in a fork and someone told me the owner of the repo had to do the setup.
- not clear what "port visibility" is about... it is private by default, do we need to make it public before accessing the URL?  at one point we got it to work, another time we got 502 Bad Gateway and couldn't figure out how to access it

**Reply by pcai** [reply]
Thanks for these - great starting place to put together an FAQ. I've been using codespaces for a few months now both personally and in some free orgs so here's my attempt to answer:
- "actions minutes" are separate from "codespaces" time. and action minutes are free/uncapped on public repositories, 2000/min/mo otherwise on free plan. but these are only charged when running CI/CD and various other optional scenarios like "codespaces prebuilds"
- codespaces billing is detailed [here](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces) but the gist is: free github personal accounts currently get 120 core-hours a month and 15gb-months of storage. back of the napkin indicates this should be good for roughly 7.5 days of 8-hr continuous usage. Codespaces storage is consumed to preserve the volume used to boot the development container; the measurement is gb-months so its completely expected for this to creep up during the month -- e.g. if my volume is .5gb then keeping the codespace around for ~30 days will consume ~15gb-months of storage (someone double check my math on this?)
- how billing is handled for forked repos is explained [here](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces#how-billing-is-handled-for-forked-repositories) but in summary, it should be billed to the personal user, but there are various scenarios that can change this behavior for organizations that need something else. but i dont think that will apply to running workshops where the optimal arrangement will be for everyone to accrue usage to their (free) personal accounts
- [this doc](https://docs.github.com/en/codespaces/developing-in-codespaces/forwarding-ports-in-your-codespace#sharing-a-port) explains when someone might want to toggle port visibility - mainly so someone else can visit their development site while the codespace is up and they have the web process running. but i would hesitate to even explain this to students for fear that it would be confused with the "deployment to an actual hosting environment" concept

**Comment by pcai** [comment]
I for one would be very happy to try one workshop where everyone is in the same environment using the same interface regardless of their os or laptop age. 

Time and again volunteers spend hours heroically working with students to "get things installed" and its very frustrating for everyone involved when the conclusion is "nothing will work on this machine" and they don't bother to come back the next day. and in the boston chapter the flowchart for contingencies is large enough that we always end up with a mix of environments, editors, etc, which makes debugging issues much more exciting than would be ideal. 

I mostly see upsides to codespaces because the web-based interface is both super accessible on almost any device with internet access, and its still fronted by vscode online. Anyone who learns it is working in a "real" editor and doesn't have to transition again if they want to go deeper.

**Reply by pcai** [reply]
To be fair, some disadvantages are:
- need reliable internet access to do literally anything
- if github has availability problems it can hard block everyone for the entire duration
